Dockerfile

```diff
@@ -1,17 +1,54 @@
-FROM ghcr.io/opensafely/base-docker
+# syntax=docker/dockerfile:1.2
+#################################################
+#
+# First build a base python image from the OS base image, so we only do this once.
+FROM ghcr.io/opensafely-core/base-docker as base-python
+COPY dependencies.txt .
+# use space efficient utility from base image
+RUN /root/docker-apt-install.sh dependencies.txt
 
-RUN apt-get update --fix-missing
+#################################################
+#
+# Next, use that to create a build image
+FROM base-python as builder
 
-# For numba
-RUN apt-get install -y llvm-10 llvm-10-dev
+# install build time dependencies 
+COPY build-dependencies.txt .
+RUN /root/docker-apt-install.sh build-dependencies.txt
 
-RUN apt-get install -y python3.8 python3.8-dev python3-pip
-RUN update-alternatives --install /usr/bin/python python /usr/bin/python3.8 1
+# install everything in venv for isolation from system python libraries
+RUN python3 -m venv /opt/venv
+ENV VIRTUAL_ENV=/opt/venv/ PATH="/opt/venv/bin:$PATH" LLVM_CONFIG=/usr/bin/llvm-config-10
 
-# Install pip requirements
-COPY requirements.txt /tmp/
-RUN LLVM_CONFIG=/usr/bin/llvm-config-10 python -m pip install --requirement /tmp/requirements.txt
+COPY requirements.txt .
+# We ensure up-to-date build tools
+# Note: the mount command does two things: 1) caches across builds to speed up
+# local development and 2) ensures the pip cache does not get committed to the
+# layer.
+RUN --mount=type=cache,target=/root/.cache \
+    python -m pip install -U pip setuptools wheel && \
+    python -m pip install --requirement requirements.txt
+
+################################################
+#
+# Finally, build the actual image from the base-python image
+FROM base-python as python
+
+# Some static metadata for this specific image
```

Comment:
Where's the best place to document our build label magic, and could we add a link to there from here?

Reply:
It's linked in the base-docker image, but doing it here is also good http://label-schema.org/rc1/

HOWEVER. 

I just noticed that its been deprecated in favor of https://github.com/opencontainers/image-spec.

So I will do follow up PRs for base-docker and this to correct the label names

Reply:
It's a bit obtuse to find, but basically it's just renaming:

https://github.com/opencontainers/image-spec/blob/master/annotations.md#back-compatibility-with-label-schema

Reply:
Isn't there also something to be mentioned about how we *use* the labels, as well as their semantics?

Reply:
I have renamed and added a link

Reply:
So, we don't use them for anything yet, as they don't exist in any images.

Once images have the metadata, there are two potential uses in mind:

1) job-runner to require an org.opensafely.action to run an image
2) the version and gitref being used in versioning and recording versions used to run specific jobs.

Basically, 1/2 of the metadata is just for public consumption, if people are looking at the images. Various registry UIs will extract that information, etc.

The other half is (build timestamps and such) are for audit/reproducablitly.

Docker image tags are ephemeral and mutable and thus unreliable. LABELs are immutable and write-once, so are a more reliable source of build information.

Reply:
It's analogous to packaging hygiene in setup.py metadata, for example.

Reply:
Thanks, this was the use-case I had in mind re. documetation.  Hard to add a comment in advance of it being used anywhere; but keen to make it easy for someone coming to this blind to have a clue about the purpose of the labels. Not sure what the answer is!